wdl-lint/src/rules/runtime_section_keys.rs

```diff
@@ -0,0 +1,355 @@
+         of any number of other keys is encouraged.
+        
+        For WDL v1.1 documents, 
+        
+        - A list of mandatory, reserved keywords will be recommended for inclusion if they are not \
+         present. Here, 'mandatory' refers to the requirement that all execution engines support \
+         this key—not that the key must be present in the `runtime` section.
+        - Optional, reserved \"hint\" keys are also permitted but not flagged when they are \
+         missing.
+        - Further, the WDL v1.1 specification deprecates the inclusion of non-reserved keys in a \
+         `runtime` section. As such, any non-reserved keys will be flagged for removal.
+         
+         For WDL v1.2 documents and later, this rule does not evaluate because `runtime` sections \
+         are deprecated."
+    }
+
+    fn tags(&self) -> crate::TagSet {
+        TagSet::new(&[Tag::Completeness, Tag::Deprecated])
+    }
+}
+
+/// A utility method to parse the recommended keys from a static set of runtime
+/// keys from either WDL v1.0 or WDL v1.1.
+fn recommended_keys<T: Deref<Target = HashSet<Key>>>(keys: T) -> HashSet<Key> {
+    keys.clone()
+        .into_iter()
+        .filter(|key| key.is_recommended())
+        .collect::<HashSet<Key>>()
+}
+
+impl Visitor for RuntimeSectionKeysRule {
+    type State = Diagnostics;
+
+    fn document(
+        &mut self,
+        _: &mut Self::State,
+        reason: wdl_ast::VisitReason,
+        document: &wdl_ast::Document,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        // Reset the visitor upon document entry.
+        *self = Default::default();
+
+        // NOTE: this rule is dependent on the version of the WDL document.
+        self.version = document
+            .version_statement()
+            .and_then(|s| s.version().as_str().parse().ok());
+    }
+
+    fn task_definition(
+        &mut self,
+        state: &mut Self::State,
+        reason: VisitReason,
+        _: &TaskDefinition,
+    ) {
+        match reason {
+            VisitReason::Enter => {
+                self.current_runtime_span = None;
+                self.keys_seen = Default::default();
+            }
+            VisitReason::Exit => {
+                // If a runtime section span has not been encountered, then
+                // presumably there won't be any place to put missing
+                // recommended runtime keys.
+                let runtime_span = match self.current_runtime_span {
+                    Some(span) => span,
+                    None => return,
+                };
+
+                if let Some(SupportedVersion::V1(minor_version)) = self.version {
+                    let recommended_keys = match minor_version {
+                        V1::Zero => recommended_keys(&V1_0_KEYS.clone()),
+                        V1::One => recommended_keys(&V1_1_KEYS.clone()),
+                        _ => Default::default(),
```

Comment:
This seems like a lot of allocations, a `clone` here, and then both a `clone` and `collect` in `recommended_keys`.

An alternative might be:

```rust
let empty = Default::default();
let recommended_keys = match minor_version {
    V1::Zero => &V1_0_KEYS,
    V1::One => &V1_1_KEYS,
    _ => &empty,
};

for key in recommended_keys.iter().filter(|key| key.is_recommended())  {
    ...
}
```

Reply:
Yeah, this is a case of not coming back to fix this: there were some quirks of `lazy_static` that I was looking at trying to avoid this, but I see now that I got a working solution and never actually came back to it. With `std::sync::OnceLock`, this appears to be much more straightforward.